File: backend/app/schemas/ai.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.utils.datetime_utils import UTCDatetime

from app.schemas.ticket import (
    RecommendedDepartment,
    TicketCategory,
    TicketPriority,
    TicketSentiment,
)


class TicketClassification(BaseModel):
    category: TicketCategory
    priority: TicketPriority
    sentiment: TicketSentiment
    need_human: bool
    summary: str = Field(min_length=1, max_length=500)
    recommended_department: RecommendedDepartment

    model_config = ConfigDict(extra="forbid")


class AIReplySource(BaseModel):
    doc_id: int
    chunk_id: int
    chunk_index: int
    content_preview: str
    score: float


class AIReplyDraftRead(BaseModel):
    id: int
    ticket_id: int
    suggestion_type: str
    source_workflow: str = "single_agent_rag"
    source_run_id: str | None = None
    suggested_content: str
    reasoning_summary: str | None
    sources_json: list[AIReplySource]
    confidence: float
    status: str
    reviewed_by: int | None = None
    reviewed_at: UTCDatetime | None = None
    final_content: str | None = None
    reject_reason: str | None = None
    created_at: UTCDatetime
    updated_at: UTCDatetime

    model_config = ConfigDict(from_attributes=True)


class AIWorkflowProcessRead(BaseModel):
    ticket: dict
    classification: dict
    knowledge_hits: list[dict]
    similar_tickets: list[dict]
    reply_suggestion: dict
    risk_check: dict
    reviewed_suggestion: dict | None = None
    review_decision: dict | None = None


class AIWorkflowPendingReviewRead(BaseModel):
    run_id: str
    thread_id: str
    status: Literal["pending_review"]
    pending_node: str
    interrupt_id: str | None = None
    ticket: dict
    classification: dict
    knowledge_hits: list[dict]
    similar_tickets: list[dict]
    draft_reply: dict
    sources: list[dict]
    confidence: float
    risk_check: dict


class AIMultiAgentPendingReviewRead(BaseModel):
    run_id: str
    thread_id: str
    status: Literal["pending_review"]
    pending_node: str
    interrupt_id: str | None = None
    ticket: dict
    supervisor_result: dict
    triage_result: dict
    knowledge_result: dict
    similar_case_result: dict
    reply_result: dict
    risk_result: dict
    workflow_result: dict
    draft_reply: dict
    sources: list[dict]
    confidence: float
    audit_trail: list[dict]


class AIMultiAgentProcessRead(BaseModel):
    ticket: dict
    supervisor_result: dict
    triage_result: dict
    knowledge_result: dict
    similar_case_result: dict
    reply_result: dict
    risk_result: dict
    workflow_result: dict
    audit_trail: list[dict] = []
    review_decision: dict | None = None
    reviewed_suggestion: dict | None = None


class AIWorkflowResumeRequest(BaseModel):
    action: Literal["approve", "edit", "reject"]
    thread_id: str | None = None
    run_id: str | None = None
    final_content: str | None = Field(default=None, min_length=1)
    reject_reason: str | None = Field(default=None, min_length=1, max_length=1000)

    @model_validator(mode="after")
    def validate_resume_payload(self) -> "AIWorkflowResumeRequest":
        if not self.thread_id and not self.run_id:
            raise ValueError("Either thread_id or run_id is required")
        if self.thread_id and self.run_id and self.thread_id != self.run_id:
            raise ValueError("thread_id and run_id must match when both are provided")
        if self.action == "edit" and not self.final_content:
            raise ValueError("final_content is required when action=edit")
        if self.action == "reject" and not self.reject_reason:
            raise ValueError("reject_reason is required when action=reject")
        return self

    def workflow_id(self) -> str:
        return self.thread_id or self.run_id or ""

File: frontend/src/api/ai.ts
import { apiClient } from "./client";
import type { TicketCategory, TicketPriority, TicketRead } from "./tickets";

export type TicketSentiment = "positive" | "neutral" | "negative" | "angry";

export type TicketClassification = {
  category: TicketCategory;
  priority: TicketPriority;
  sentiment: TicketSentiment;
  need_human: boolean;
  summary: string;
  recommended_department: string;
};

export type AIReplySource = {
  doc_id: number;
  chunk_id: number;
  chunk_index: number;
  content_preview: string;
  score: number;
};

export type AIReplyDraftRead = {
  id: number;
  ticket_id: number;
  suggestion_type: string;
  source_workflow: "single_agent_rag" | "single_agent_workflow" | "multi_agent" | "manual" | string;
  source_run_id: string | null;
  suggested_content: string;
  reasoning_summary: string | null;
  sources_json: AIReplySource[];
  confidence: number;
  status: string;
  reviewed_by: number | null;
  reviewed_at: string | null;
  final_content: string | null;
  reject_reason: string | null;
  created_at: string;
  updated_at: string;
};

export function normalizeSourceWorkflow(value: string | undefined | null): string {
  if (!value || value === "single_agent") {
    return "single_agent_rag";
  }
  if (value === "workflow") {
    return "single_agent_workflow";
  }
  return value;
}

export type SuggestionApprovePayload = {
  final_content?: string;
};

export type SuggestionEditPayload = {
  final_content: string;
};

export type SuggestionRejectPayload = {
  reject_reason: string;
};

export type AgentAuditTrailItem = {
  agent_name: string;
  action: string;
  input_summary: string;
  output_summary: string;
  status: string;
  timestamp: string;
};

export type MultiAgentSupervisorResult = {
  agent_name: string;
  workflow_mode: string;
  planned_agents: string[];
  requires_human_review: boolean;
  summary: string;
};

export type MultiAgentTriageResult = {
  agent_name: string;
  classification: TicketClassification;
};

export type MultiAgentKnowledgeHit = AIReplySource & {
  embedding_id?: string | null;
};

export type MultiAgentKnowledgeResult = {
  agent_name: string;
  query: string;
  confidence: number;
  low_confidence_reason: string | null;
  hits: MultiAgentKnowledgeHit[];
};

export type MultiAgentSimilarCaseTicket = {
  ticket_id: number;
  title: string;
  similarity: number;
  resolution: string;
  content_preview?: string;
  [key: string]: unknown;
};

export type MultiAgentSimilarCaseResult = {
  agent_name: string;
  similar_tickets: MultiAgentSimilarCaseTicket[];
  historical_summary: string;
};

export type MultiAgentReplyResult = {
  agent_name: string;
  supplemental_context: string;
  reply_suggestion: AIReplyDraftRead;
};

export type MultiAgentRiskCheck = {
  risk_level: "low" | "medium" | "high" | string;
  requires_human_review: boolean;
  reasons: string[];
};

export type MultiAgentRiskResult = {
  agent_name: string;
  risk_check: MultiAgentRiskCheck;
};

export type MultiAgentWorkflowResult = {
  agent_name: string;
  next_status: string;
  assign_to_department: string;
  next_action: string;
  internal_note: string;
  updated_ticket: TicketRead;
};

export type AIMultiAgentPendingReviewRead = {
  run_id: string;
  thread_id: string;
  status: "pending_review";
  pending_node: string;
  interrupt_id: string | null;
  ticket: TicketRead;
  supervisor_result: MultiAgentSupervisorResult;
  triage_result: MultiAgentTriageResult;
  knowledge_result: MultiAgentKnowledgeResult;
  similar_case_result: MultiAgentSimilarCaseResult;
  reply_result: MultiAgentReplyResult;
  risk_result: MultiAgentRiskResult;
  workflow_result: MultiAgentWorkflowResult;
  draft_reply: AIReplyDraftRead;
  sources: AIReplySource[];
  confidence: number;
  audit_trail: AgentAuditTrailItem[];
};

export type MultiAgentResumePayload = {
  action: "approve" | "edit" | "reject";
  thread_id?: string;
  run_id?: string;
  final_content?: string;
  reject_reason?: string;
};

export type AIMultiAgentProcessRead = {
  ticket: Record<string, unknown>;
  supervisor_result: MultiAgentSupervisorResult;
  triage_result: MultiAgentTriageResult;
  knowledge_result: MultiAgentKnowledgeResult;
  similar_case_result: MultiAgentSimilarCaseResult;
  reply_result: MultiAgentReplyResult;
  risk_result: MultiAgentRiskResult;
  workflow_result: MultiAgentWorkflowResult;
  audit_trail: AgentAuditTrailItem[];
  review_decision: Record<string, unknown> | null;
  reviewed_suggestion: AIReplyDraftRead | null;
};

export type AgentRunLogRead = {
  id: number;
  ticket_id: number;
  run_id: string;
  run_type: "single_agent_rag" | "single_agent_workflow" | "multi_agent" | "workflow" | string;
  status: string;
  input_json: Record<string, unknown>;
  output_json: Record<string, unknown>;
  audit_trail_json: AgentAuditTrailItem[];
  error_message: string | null;
  created_by: number | null;
  created_at: string;
  updated_at: string;
};

export async function classifyTicketAI(ticketId: number) {
  const response = await apiClient.post<TicketClassification>(`/ai/tickets/${ticketId}/classify`);
  return response.data;
}

export async function generateTicketReply(ticketId: number) {
  const response = await apiClient.post<AIReplyDraftRead>(`/ai/tickets/${ticketId}/generate-reply`);
  return response.data;
}

export async function listTicketSuggestions(ticketId: number) {
  const response = await apiClient.get<AIReplyDraftRead[]>(`/ai/tickets/${ticketId}/suggestions`);
  return response.data;
}

export async function approveSuggestion(suggestionId: number, payload: SuggestionApprovePayload = {}) {
  const response = await apiClient.post<AIReplyDraftRead>(`/reviews/${suggestionId}/approve`, payload);
  return response.data;
}

export async function editSuggestion(suggestionId: number, payload: SuggestionEditPayload) {
  const response = await apiClient.post<AIReplyDraftRead>(`/reviews/${suggestionId}/edit`, payload);
  return response.data;
}

export async function rejectSuggestion(suggestionId: number, payload: SuggestionRejectPayload) {
  const response = await apiClient.post<AIReplyDraftRead>(`/reviews/${suggestionId}/reject`, payload);
  return response.data;
}

export async function startMultiAgentProcess(ticketId: number) {
  const response = await apiClient.post<AIMultiAgentPendingReviewRead>(
    `/ai/tickets/${ticketId}/multi-agent-process/start`,
  );
  return response.data;
}

export async function resumeMultiAgentProcess(
  ticketId: number,
  payload: MultiAgentResumePayload,
) {
  const response = await apiClient.post<AIMultiAgentProcessRead>(
    `/ai/tickets/${ticketId}/multi-agent-process/resume`,
    payload,
  );
  return response.data;
}

export async function listTicketAgentRuns(ticketId: number) {
  const response = await apiClient.get<AgentRunLogRead[]>(`/ai/tickets/${ticketId}/agent-runs`);
  return response.data;
}

export type AgentRunLogPage = {
  items: AgentRunLogRead[];
  total: number;
  limit: number;
  offset: number;
};

export type AgentRunLogPageParams = {
  run_type?: string;
  status?: string;
  limit?: number;
  offset?: number;
};

export async function listTicketAgentRunsPage(
  ticketId: number,
  params: AgentRunLogPageParams = {},
) {
  const response = await apiClient.get<AgentRunLogPage>(
    `/ai/tickets/${ticketId}/agent-runs/page`,
    { params },
  );
  return response.data;
}

export type LatestAgentRunsByType = {
  single_agent_rag: AgentRunLogRead | null;
  single_agent_workflow: AgentRunLogRead | null;
  multi_agent: AgentRunLogRead | null;
};

export async function getLatestTicketAgentRunsByType(
  ticketId: number,
): Promise<LatestAgentRunsByType> {
  const response = await apiClient.get<LatestAgentRunsByType>(
    `/ai/tickets/${ticketId}/agent-runs/latest-by-type`,
  );
  return response.data;
}

export type AIWorkflowPendingReviewRead = {
  run_id: string;
  thread_id: string;
  status: "pending_review";
  pending_node: string;
  interrupt_id: string | null;
  ticket: TicketRead;
  classification: TicketClassification;
  knowledge_hits: Record<string, unknown>[];
  similar_tickets: Record<string, unknown>[];
  draft_reply: AIReplyDraftRead;
  sources: AIReplySource[];
  confidence: number;
  risk_check: Record<string, unknown>;
};

export type AIWorkflowProcessRead = {
  ticket: Record<string, unknown>;
  classification: Record<string, unknown>;
  knowledge_hits: Record<string, unknown>[];
  similar_tickets: Record<string, unknown>[];
  reply_suggestion: Record<string, unknown>;
  risk_check: Record<string, unknown>;
  reviewed_suggestion: AIReplyDraftRead | null;
  review_decision: Record<string, unknown> | null;
};

export type AIWorkflowResumePayload = {
  action: "approve" | "edit" | "reject";
  thread_id?: string;
  run_id?: string;
  final_content?: string;
  reject_reason?: string;
};

export async function startSingleAgentProcess(ticketId: number) {
  const response = await apiClient.post<AIWorkflowPendingReviewRead>(
    `/ai/tickets/${ticketId}/process/start`,
  );
  return response.data;
}

export async function resumeSingleAgentProcess(
  ticketId: number,
  payload: AIWorkflowResumePayload,
) {
  const response = await apiClient.post<AIWorkflowProcessRead>(
    `/ai/tickets/${ticketId}/process/resume`,
    payload,
  );
  return response.data;
}

export async function listReviewedSuggestions(ticketId: number) {
  const response = await apiClient.get<AIReplyDraftRead[]>(
    `/ai/tickets/${ticketId}/reviewed-suggestions`,
  );
  return response.data;
}
